Bug fixes and improvements

Alex-Kent opened the pull request:
This patch does the following:

- Watching multiple files/layers is now working properly
- Support added for file names that are URL-encoded (including those with options appended); this includes things like delimited text files (CSV and the like).

The code is reasonably-well documented, let me know if you have any questions.

Alex-Kent committed:
• Watching multiple files/layers is now working properly 
• Support added for file names that are URL-encoded (including those with options appended)

Alex-Kent commented on reloader.py:
```diff
@@ -11,6 +11,17 @@
+ o Support added for file names that are URL-encoded (including those with 
+   options appended)
+
+ ǒ 2025 Alexander Hajnal
```
Should read `© 2025 Alexander Hajnal`

Alex-Kent commented on reloader.py:
```diff
@@ -11,6 +11,17 @@
  ***************************************************************************/
+ 
+/***************************************************************************
+ Updated 2025-020-11:
```
Should read `Updated 2025-02-11:`

Alex-Kent commented on README.md:
```diff
@@ -6,4 +6,10 @@ reopen also updates the extent. Also provides an option to watch layer
+do this open the Properties for the layer that is Joined to the delimted file 
+(not the delimited file's layer) and select that file's layer in the list under 
+Dependencies.  After doing this the joined layer should update automatically 
+when the delimited file is watched for changes.
```
One additional setting needs to be made; this should read:
```
In addition to e.g. shapefiles, delimited text files can also be monitored.  To 
do this open the Properties for the layer that is Joined to the delimted file 
(not the delimited file's layer) and select that file's layer in the list under 
Dependencies, then under Joins edit the join layer and make sure Cache join 
layer in memory is not selected.  After doing this the joined layer should 
update automatically when the watched delimited file changes.
```

Alex-Kent commented:
I can submit a new PR if you'd like.

evetion commented:
Thanks for the PR! I'll take a look tomorrow. In the meantime, can you explain a bit on what things didn't work before? Either edited in the PR description above, or a new issue?

> I can submit a new PR if you'd like.

How familiar are you with git? You should be able to just commit to the main branch of your fork where you did this work to update this PR.

Alex-Kent commented:
Why don't I submit a bug report for the two things that were changed.

As for Git, I'm somewhat proficient but I'm not hugely familiar with how Github handles things.  I'll try pushing an update to my repo and see if it propagates here.

Alex-Kent committed:
Documentation corrections

evetion commented on README.md:
```diff
@@ -6,4 +6,11 @@ reopen also updates the extent. Also provides an option to watch layer
 for changes to local file paths and reloads them automatically.
 Useful when you're editing the underlying data on the fly in some other program.
 
+In addition to e.g. shapefiles, delimited text files can also be monitored.  To 
```
I think you meant to say that joined layers can be monitored as well?

Alex-Kent replied:
The delimited text file is the one being watched.  The layers rendered in the GUI are updated as a side effect of the the delimited file layer being refreshed (due to the latter being listed as a dependency for the rendered layer), i.e. the change notification propagates from the text file layer.  The layer being rendered (e.g. a static shapefile) does not need to be watched.

Alex-Kent replied:
> I'm also not sure what didn't work in the old code, did the layers not update in some cases?

On my system the wrong layer is updated.  This appears to be due to a Python bug on some systems.  I'll be submitting a bug report detailing the issue and workaround shortly.

Alex-Kent replied:
> pass the path, but you don't need it in the function (only for logging)

When I use the following, the value that callback receives via its `layer_id` argument is that of `path`, not that of `layer_id`:

```python
def reload(layer_id):
…
watcher.fileChanged.connect(lambda layer_id=layer_id: reload(layer_id))
```

When changed to the following both `layer_id` and `path` are correctly passed:
```python
def reload(path, layer_id):
…
watcher.fileChanged.connect(lambda path=path, layer_id=layer_id: reload(path, layer_id))
```

However, order matters as the following has the value of `path` passed for both `path` and `layer_id`:
```python
def reload(layer_id, path):
…
watcher.fileChanged.connect(lambda layer_id=layer_id, path=path: reload(layer_id, path))
```

No, this doesn't make sense.

evetion replied:
The watcher is a [pure Qt thing](https://doc.qt.io/qt-6/qfilesystemwatcher.html#fileChanged) (not a QGIS thing), and since it watches for changes in a file, it will only pass the filepath through the connect. It doesn't know about layer_id.

With python rules about arguments, it will be set to the first argument (whatever the name), and apparently also to the first named argument called path. In all cases layer_id comes from the enclosing code (which might probably be buggy for watching multiple layers at the same time).

evetion replied:
And I think the bug is about late binding, consider the following code:

```python
d = {}
for i in [1,2,3]:
    def test():
        return(i)
    d[i] = test
d
{1: <function test at 0x10549a340>, 2: <function test at 0x10549a3e0>, 3: <function test at 0x10549a2a0>}
d[1]()
3
```

Even though the funcion tests are different, they all bounded to i at the end of the loop (late), capturing 3. Reloader has the same bug, so if multiple layers are selected, only the last layer will be reloaded (multiple times).

The fix is to provide a default argument like so `i=i`:
```
d = {}
for i in [1,2,3]:
    def test(i=i):
        return(i)
    d[i] = test
d[1]()
1
```

So the code should change to
```
def reload(path, layer_id=layer_id):
...
watcher.fileChanged.connect(reload)
```

Alex-Kent replied:
> In all cases layer_id comes from the enclosing code (which might probably be buggy for watching multiple layers at the same time).

Apparently the variable values passed to callbacks are those at the time the callback is called, not those at the time that the callback is created [my previous understanding was that the latter was the case].  I found [a discussion of this](http://jceipek.com/Olin-Coding-Tutorials/) where the author explores the behavior in depth.  They used a factory function to snapshot the variable values whilst I used a lambda to achieve a similar result [not the same behavior, but adequate for this application].

[I'm very much a neophyte when it comes to Python so if my statements appear to contradict one-another it's because I've learned a new language nuance.]

Alex-Kent replied:
Regarding `watcher`'s behavior, I was not aware that it modifies the values passed to the callback function's arguments.  That would explain the behavior I'm seeing.  (I had assumed that the callback was called verbatim.)

Alex-Kent replied:
> I think the bug is about late binding

I concur.  It seems there are at least three ways to solve this issue:

1. What you propose:
```python
def reload(path, layer=layer):
    layer.reload()
    layer.triggerRepaint()
watcher.fileChanged.connect(reload)
```
2. My code:
```python
def reload(path, layer):
    layer.reload()
    layer.triggerRepaint()
watcher.fileChanged.connect(lambda path=path, layer=layer: reload(path, layer))
```
3. Use a factory function (as discussed in the link above), perhaps:
```python
def reload_factory(layer):
    def reload(path):
        layer.reload()
        layer.triggerRepaint()
    return reload
watcher.fileChanged.connect(reload_factory(layer))
```

I've tested all three options and they all work correctly.

I'm partial to either option 1 or option 2.  Your code is cleaner than mine but mine is more explicit about the state being captured (writing it as `watcher.fileChanged.connect(lambda p='overwritten', l=layer: reload(p, l))` would be even more explicit about what's going on).

evetion replied:
I like my (1) option, because the lambda (2) removes the distinction between what's been passed from the signal, and what we store from the local scope. That said, it still needs an inline comment that explains exactly that, so we don't make the same mistake in the future.

Alex-Kent replied:
Just saw your comment.  I think you're saying that by putting the assignment in the function declaration it makes it clearer that the value is being set using early binding, am I understanding you correctly?  I alluded to the early vs. late binding issue with a comment immediately preceding the callback function's connection to the watcher (see https://github.com/evetion/Reloader/pull/1/commits/a0b86adb2668e86be63e557efc76e3448ec0c61f), this could also be noted in a comment alongside the function definition.  Would you like me to change the code to use your solution?  Regardless I'll update the comments to describe the binding behavior more explicitly.

evetion commented on reloader.py:
```diff
@@ -232,22 +254,71 @@ def watch(self):
+                
+                # For e.g. a delimited text layer, the raw path is URL encoded 
+                # and has options appended; this needs to be cleaned up.
+                if re.match("^file://", path):
+                    # Strip the prefix
+                    path = re.sub("^file://", "", path)
+                    # Strip the options off the end
+                    path = re.sub("\?.*$", "", path)
+                    # Decode the filename into UTF-8
+                    path = urllib.parse.unquote(path)
```
Could you move this into a separate function that we can call here? That makes this part (more) easily testable.

Alex-Kent replied:
Sure thing.  I presume you'd like the conditional to be within the function?  For the call how about `path = extract_path_from_uri(path)`?

I have some other tweaks in mind; I'll code an implementation and you can let me know what you think.

Alex-Kent replied:
Implemented in https://github.com/evetion/Reloader/pull/1/commits/3bd73e44ac279874b9ad4219b88c6bdc11cbb713

evetion commented on reloader.py:
```diff
@@ -232,22 +254,71 @@ def watch(self):
+                        # if layer_id is made the first argument then both values 
+                        # are set to the path!
+                        
+                        QgsMessageLog.logMessage(f"Reloading layer", 'Reloader')
+                        
+                        # Get the layer object with the passed ID
+                        # Things break if "layer" is used as the (left-side) variable name
+                        l = QgsProject.instance().mapLayer(layer_id)
+                        
+                        QgsMessageLog.logMessage( "    ID:    " + l.id(),   'Reloader')
+                        QgsMessageLog.logMessage( "    Name:  " + l.name(), 'Reloader')
+                        QgsMessageLog.logMessage( "    Path:  " + path,     'Reloader')
```
This will notify the user I believe, might be good to set `notifyUser=False` in logMessage. Also, could we merge these messages into a single string, or remove them altogether (or set them to debug level?).

Alex-Kent replied:
As these aren't required for the code to function I'd lean towards removing them, setting them to debug level or commenting them out.  Regardless I'll merge the messages, explicitly set their level to `Info` (the lowest level) and set `notifyUser` to `false`.

Alex-Kent replied:
Implemented in https://github.com/evetion/Reloader/pull/1/commits/0b9ffeaa0845a7ae6494aa5f4db626d1c0354885

evetion reviewed (commented):
I'm not sure about the change to the monitoring. You have added a lot of warnings about variable naming/aliasing, and pass the path, but you don't need it in the function (only for logging). So the main change is to use the layer_id, but that is not passed by the signal, and still cached from the code above it. The signal only passes the path I guess, and that might set it to both arguments?

I'm also not sure what didn't work in the old code, did the layers not update in some cases?

Let me know if you want me to work on the PR.

Alex-Kent commented:
I'll be submitting a second bug report shortly regarding the changes to to callback function.  In a nutshell it's a workaround for what appears to be a Python bug (likely system-dependent).

Alex-Kent committed:
Moved URI decoding code into a function (per https://github.com/evetion/Reloader/pull/1#discussion_r1951525938)

Alex-Kent committed:
Adjusted logging calls (per https://github.com/evetion/Reloader/pull/1#discussion_r1951534519)

Alex-Kent committed:
Cleaned up callback code (see https://github.com/evetion/Reloader/pull/1#discussion_r1952628513 and following posts)

Alex-Kent committed:
Cleaned up callback code (see https://github.com/evetion/Reloader/pull/1#discussion_r1952496480 and following posts)

Alex-Kent committed:
Merge branch 'main' of git@github.com:Alex-Kent/Reloader.git

Alex-Kent committed:
Updated callback to reinstall watch if watcher removed it (resolves https://github.com/evetion/Reloader/issues/4)

Alex-Kent committed:
Tweaked code order in callback (for clarity)

Alex-Kent commented:
All of the changes that we've been discussing plus the fix for https://github.com/evetion/Reloader/issues/4 have been implemented and pushed to Github.  For installing the callback I used my implementation (a lambda); as this is your project though it's ultimately your decision as to which method to use for this.  Currently I'm passively testing the code (it's being triggered every minute or so) and I haven't seen any issues so far.

Alex-Kent committed:
• Changed watcher callback to pass parameters via its definition
• Improved comments
• Changed callback function name since a method with the same name also exists (this has no effect on the code's function)

Alex-Kent commented:
The most recent push (https://github.com/evetion/Reloader/pull/1/commits/14ec5f826fab1a2f512b6a4cf2bfde3ebae08671) covers everything that we've been discussing.  This code has been tested for the past 24 hours with multiple file-backed layers being watched and no issues have been seen.  In my opinion this code should be ready to be merged.

evetion commented:
Thanks for the finding the bugs, reporting and fixing them! I will make some minor changes and make a release this weekend.

Alex-Kent commented:
> Thanks for the finding the bugs, reporting and fixing them!

¡De nada!  And thank you for writing such a useful plugin!

> I will make some minor changes

There's always something missed: lines 66 and 67 aren't needed and can be deleted (that's the `from qgis.core import QgsProject` statement and its comment).  I've updated my code and am passively testing to ensure it can be safely removed (which it should as it was only used for retrieving the layer by its ID in my initial push).  This change was pushed to my copy of the repo after you merged the code (https://github.com/evetion/Reloader/commit/4e49c7ab4199ebdcade3b4fefd04276ebc50c8a8).

evetion commented:
I've removed some lines, ran an autoformatter, added a test and some more in https://github.com/evetion/Reloader/commit/1577f350dfbad3b3b28907f7f9fffb95040931de. Most impactful, I've shortened a check you did for re-adding the watch at https://github.com/evetion/Reloader/commit/1577f350dfbad3b3b28907f7f9fffb95040931de#diff-859e6c5392dbec182d14bd578888ecdd0c6c29b9007f5abce8d7654f541b2418R356. Can you confirm it still works as it should?

edit: See, it was good to check, I think I inverted the condition there.

Alex-Kent commented:
> Can you confirm it still works as it should?

I read through the [diff](https://github.com/evetion/Reloader/commit/1577f350dfbad3b3b28907f7f9fffb95040931de?diff=split#diff-859e6c5392dbec182d14bd578888ecdd0c6c29b9007f5abce8d7654f541b2418R337-R339) and saw a few things, mostly minor (stylistic), one or two that might be issues.  I'll pull the code and test it then annotate the aforementioned diff.

Alex-Kent commented:
~~Correction: I'll annotate the [diff](https://github.com/evetion/Reloader/compare/5083128...1f35f1689e01fd976e764c416afaabe05464ad97) between your merge of my pull request (https://github.com/evetion/Reloader/commit/50831287c88e77fa380b9b5732b8d55bdf8060a8) and your most recent push (https://github.com/evetion/Reloader/commit/1f35f1689e01fd976e764c416afaabe05464ad97).~~  [edit: Or not, Github won't let me annotate that diff.]

Alex-Kent commented:
How are you running `test.py`?

Alex-Kent commented:
I added comments to https://github.com/evetion/Reloader/commit/1f35f1689e01fd976e764c416afaabe05464ad97 and https://github.com/evetion/Reloader/commit/1577f350dfbad3b3b28907f7f9fffb95040931de

The current code is running fine on my system (it's been running with watched file updates [for multiple files] every minute or two for the past 4 hours).

evetion commented:
Great, thanks for testing! I'll make a release tomorrow.

Alex-Kent commented:
There are a few suggested changes in the comments I made.

~~Amongst other things, can you test this on Windows or would you like me to install QGIS in a VM and test it there?~~ [edit: I did the latter.]  To be specific, on Windows `/` characters in `file://` URIs may need to be changed to `\`.

Alex-Kent commented:
I was right, `file://` URI aren't handled correctly on Windows (there's even an odd added quirk).  Fixing it looks trivial.  I'll submit a bug report and a new PR to fix the issue.
